sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -57,26 +90,135 @@ case class GpuNot(child: Expression) extends CudfUnaryExpression
   }
 }
 
-case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryOperator with Predicate {
+abstract class CudfBinaryPredicateWithSideEffect extends CudfBinaryOperator
+  with Predicate with GpuCVPredicateHelper {
+  
   override def inputType: AbstractDataType = BooleanType
 
+  def columnarEvalWithSideEffects(batch: ColumnarBatch): Any
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val rightExpr = right.asInstanceOf[GpuExpression]
+
+    if (rightExpr.hasSideEffects) {
+      columnarEvalWithSideEffects(batch)
+    } else {
+      super.columnarEval(batch)
+    }
+  }
+}
+
+case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryPredicateWithSideEffect {
+  
   override def symbol: String = "&&"
 
   override def sqlOperator: String = "AND"
 
   override def binaryOp: BinaryOp = BinaryOp.NULL_LOGICAL_AND
   override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.NULL_LOGICAL_AND)
+
+
+  /**
+   * When computing logical expressions on the CPU, the true and false
+   * expressions are evaluated lazily, meaning that the RHS expression
+   * of logical-AND is not evaluated when LHS is False. For logical-OR,
+   * RHS is not evaluated when LHS is True.
+   * This is important in the case where the expressions can have
+   * side-effects, such as throwing exceptions for invalid inputs.
+   *
+   * This method performs lazy evaluation on the GPU by first filtering
+   * the input batch a  where the LHS predicate is True.
+   * The RHS predicate is evaluated against these batches and then the
+   * results are combined back into a single batch using the gather
+   * algorithm.
+   */
+  override def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val leftExpr = left.asInstanceOf[GpuExpression]
+    val rightExpr = right.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(GpuExpressionsUtils.columnarEvalToColumn(leftExpr, batch)) { lhsBool =>
+        if (isAllFalse(lhsBool)) {
+          withResource(GpuScalar.from(false, dataType)) { falseScalar =>
+            GpuColumnVector.from(falseScalar, lhsBool.getRowCount.toInt, BooleanType)
+          }
+        } else {
+          // replace nulll values
+          val lhsNoNulls = withResource(Scalar.fromBool(true)) { trueScalar =>
+            lhsBool.getBase.replaceNulls(trueScalar)
+          }
+          withResource(lhsNoNulls) { _ =>
+            val rEval = withResource(filterBatch(tbl, lhsNoNulls, colTypes)) { leftTrueBatch =>
+              rightExpr.columnarEval(leftTrueBatch)
+            }
+            val finalRet = withResourceIfAllowed(rEval) { _ =>
+              (rEval) match {
```

Comment:
The parentheses here are creating a tuple containing `rEval` and are not required. They will need removing from the `case` lines as well.

Reply:
Done!